### PR TITLE
[SPARK-37512][PYTHON] Support TimedeltaIndex creation (from Series/Index) and TimedeltaIndex.astype

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Support TimedeltaIndex creation given a timedelta Series/Index.

`astype` is also supported naturally in this PR.

### Why are the changes needed?
Follow panda's behavior.

### Does this PR introduce _any_ user-facing change?
Yes.

#### TimedeltaIndex creation (from Series/Index)
```py
>>> idx = ps.TimedeltaIndex([timedelta(1), timedelta(microseconds=2)])
>>> s = ps.Series([timedelta(1), timedelta(microseconds=2)], index=[10, 20])

## FROM ## 
>>> ps.TimedeltaIndex(idx)
Traceback (most recent call last):
...
NotImplementedError: Create a TimedeltaIndex from Index/Series is not supported

>>> ps.TimedeltaIndex(s)
Traceback (most recent call last):
...
NotImplementedError: Create a TimedeltaIndex from Index/Series is not supported

## TO ##
>>> ps.TimedeltaIndex(idx)
TimedeltaIndex(['1 days 00:00:00', '0 days 00:00:00.000002'], dtype='timedelta64[ns]', freq=None)

>>> ps.TimedeltaIndex(s)
TimedeltaIndex(['1 days 00:00:00', '0 days 00:00:00.000002'], dtype='timedelta64[ns]', freq=None)

```
#### TimedeltaIndex.astype
```py
>>> psidx = ps.TimedeltaIndex([timedelta(1)])
>>> psidx.astype(str)
Index(['INTERVAL '1 00:00:00' DAY TO SECOND'], dtype='object')
>>> psidx.astype(int)
Int64Index([86400], dtype='int64')
>>> psidx.astype('category')
CategoricalIndex(['1 days'], categories=[1 days 00:00:00], ordered=False, dtype='category')
```

### How was this patch tested?
Unit tests.

### CODE REVIEW

**Comment by xinrong-databricks** [comment]
The test failure is irrelevant to the PR
```py
======================================================================
ERROR [2.161s]: test_reuse_worker_of_parallelize_range (pyspark.tests.test_worker.WorkerReuseTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/tests/test_worker.py", line 195, in test_reuse_worker_of_parallelize_range
    self.assertTrue(pid in previous_pids)
AssertionError: False is not true

```

**Comment by xinrong-databricks** [comment]
FYI @ueshin @HyukjinKwon @itholic

File: python/pyspark/pandas/data_type_ops/timedelta_ops.py [hunk 1]
```diff
@@ -26,3 +42,19 @@ class TimedeltaOps(DataTypeOps):
...
     def pretty_name(self) -> str:
         return "timedelta"
+
+    def astype(self, index_ops: IndexOpsLike, dtype: Union[str, type, Dtype]) -> IndexOpsLike:
+        dtype, spark_type = pandas_on_spark_type(dtype)
+
+        if isinstance(dtype, CategoricalDtype):
+            return _as_categorical_type(index_ops, dtype, spark_type)
+        elif isinstance(spark_type, BooleanType):
+            raise TypeError("cannot astype a datetimelike from [timedelta64[ns]] to [bool]")
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/data_type_ops/timedelta_ops.py:52]: should we use `dtype` for `timedelta64[ns]`?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: python/pyspark/pandas/data_type_ops/timedelta_ops.py (hunk 1 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/data_type_ops/timedelta_ops.py:52]: Certainly, will adjust it in https://github.com/apache/spark/pull/34787, thanks!

**Pull request closed without merging**